src/test/scala/edu/ie3/simona/model/participant/StorageModelSpec.scala

```diff
@@ -0,0 +1,418 @@
+      val storageModel = buildStorageModel(Some(0.5d))
+      val startTick = 3600L
+      val data = StorageModel.StorageRelevantData(startTick + 1)
+
+      val testCases = Table(
+        ("lastStored", "pRef", "pMin", "pMax"),
+        // completely empty
+        (0.0, 10.0, 0.0, 10.0),
+        // below margin of ref power target
+        (49.9974, 10.0, -10.0, 10.0),
+        // within margin below ref power target
+        (49.9976, 0.0, -10.0, 10.0),
+        // exactly at ref power target
+        (50.0, 0.0, -10.0, 10.0),
+        // within margin above ref power target
+        (50.0030, 0.0, -10.0, 10.0),
+        // above margin of ref power target
+        (50.0031, -10.0, -10.0, 10.0),
+        // at mid-level charge
+        (60.0, -10.0, -10.0, 10.0),
+        // fully charged
+        (100.0, -10.0, -10.0, 0.0),
+      )
+
+      forAll(testCases) {
+        (lastStored: Double, pRef: Double, pMin: Double, pMax: Double) =>
+          val oldState = StorageModel.StorageState(
+            KilowattHours(lastStored),
+            Kilowatts(0d),
+            startTick,
+          )
+
+          val result = storageModel
+            .determineFlexOptions(data, oldState)
+            .asInstanceOf[ProvideMinMaxFlexOptions]
+
+          result.ref should approximate(Kilowatts(pRef))
+          result.min should approximate(Kilowatts(pMin))
+          result.max should approximate(Kilowatts(pMax))
+      }
+    }
+
+    "Handle controlled power change" in {
+      val storageModel = buildStorageModel()
+      val startTick = 3600L
+      val data = StorageModel.StorageRelevantData(startTick + 1)
+      val testCases = Table(
+        (
+          "lastStored",
+          "setPower",
+          "expPower",
+          "expActiveNext",
+          "expScheduled",
+          "expDelta",
+        ),
+        // no power
+        (0.0, 0.0, 0.0, false, false, 0),
+        (50.0, 0.0, 0.0, false, false, 0),
+        (100.0, 0.0, 0.0, false, false, 0),
+        // charging on empty
+        (0.0, 1.0, 1.0, true, true, (100 * 3600 / 0.9).toInt),
+        (0.0, 2.5, 2.5, true, true, (40 * 3600 / 0.9).toInt),
+        (0.0, 5.0, 5.0, true, true, (20 * 3600 / 0.9).toInt),
+        (0.0, 10.0, 10.0, true, true, (10 * 3600 / 0.9).toInt),
+        // charging on half full
+        (50.0, 5.0, 5.0, false, true, (10 * 3600 / 0.9).toInt),
+        (50.0, 10.0, 10.0, false, true, (5 * 3600 / 0.9).toInt),
+        // discharging on half full
+        (50.0, -4.5, -4.5, false, true, 10 * 3600),
+        (50.0, -9.0, -9.0, false, true, 5 * 3600),
+        // discharging on full
+        (100.0, -4.5, -4.5, true, true, 20 * 3600),
+        (100.0, -9.0, -9.0, true, true, 10 * 3600),
+      )
+
+      forAll(testCases) {
+        (
+            lastStored: Double,
+            setPower: Double,
+            expPower: Double,
+            expActiveNext: Boolean,
+            expScheduled: Boolean,
+            expDelta: Int,
```

Comment:
Maybe we need to take another look together because when I remove .toInt and set expDelta to double, it wants an AnyVal."required: (Double, Double, Double, Boolean, Boolean, AnyVal) => ?" Maybe because changesAtTick need a Long?
```scala
final case class FlexChangeIndicator(
    changesAtNextActivation: Boolean = false,
    changesAtTick: Option[Long] = None,
)``

Reply:
I’ll try something else again, maybe the map conversion in the Groovy test was the issue.